wcfsetup/install/files/lib/system/session/SessionHandler.class.php

```diff
@@ -405,46 +370,114 @@ public function getUser() {
 	}
 	
 	/**
-	 * Tries to read existing session identified by the given session id.
-	 * 
-	 * @param	string		$sessionID
+	 * Tries to read existing session identified by the given session id. Returns whether
+	 * a session could be found.
 	 */
-	protected function getExistingSession($sessionID) {
-		$this->session = new $this->sessionClassName($sessionID);
-		if (!$this->session->sessionID) {
-			$this->session = null;
-			return;
+	protected function getExistingSession(string $sessionID): bool {
+		$sql = "SELECT	*
+			FROM	wcf".WCF_N."_".($this->isACP ? 'acp' : 'user')."_session
+			WHERE	sessionID = ?";
+		$statement = WCF::getDB()->prepareStatement($sql);
+		$statement->execute([
+			$sessionID
+		]);
+		$row = $statement->fetchSingleRow();
+		
+		if (!$row) {
+			return false;
+		}
+		
+		// Check whether the session technically already expired.
+		$lifetime = ($this->isACP ? self::ACP_SESSION_LIFETIME :
+			($row['userID'] ? self::USER_SESSION_LIFETIME :
+			self::GUEST_SESSION_LIFETIME));
+		if ($row['lastActivityTime'] < (TIME_NOW - $lifetime)) {
+			return false;
+		}
+		
+		$this->sessionID = $sessionID;
+		$this->user = new User($row['userID']);
+		$this->variables = unserialize($row['sessionVariables']);
+		
+		$sql = "UPDATE	wcf".WCF_N."_".($this->isACP ? 'acp' : 'user')."_session
+			SET	ipAddress = ?,
+				userAgent = ?,
+				lastActivityTime = ?
+			WHERE	sessionID = ?";
+		$statement = WCF::getDB()->prepareStatement($sql);
+		$statement->execute([
+			UserUtil::getIpAddress(),
+			UserUtil::getUserAgent(),
+			TIME_NOW,
+			$this->sessionID,
+		]);
+		
+		// Refresh cookie.
+		if ($this->user->userID) {
+			HeaderUtil::setCookie(($this->isACP ? 'acp' : 'user')."_session", $this->sessionID, TIME_NOW + 86400 * 14);
```

Comment:
This appears to send a cookie with every response?

Reply:
Yes, because with long lived sessions of 14 days after the last activity we need to take care that the cookie also lives for these 14 days. Storing some timestamp to only update the cookie once per hour or so feels fragile.

Reply:
That's quite extreme and I have never seen a site behave like this. Why do we even care, just let the cookie live until eternity. The session id is either valid or not, it makes no difference what the browser presents to us.

Reply:
> That's quite extreme and I have never seen a site behave like this.

GitHub sends an updated session cookie with every request. Laravel does as well (by default?).

> Why do we even care, just let the cookie live until eternity.

I wouldn't be surprised if browsers start to restrict the maximum age of a cookie. IMO it's also questionable to store an “eternity cookie” on a users device when the user might not visit the community any longer. Also even an eternity cookie will expire one day, leading to surprise logouts. Personally I'm always annoyed when I'm suddenly logged out, because one year passed since I logged in last and my autologin cookies expired. This is especially annoying on my phone.

Reply:
Restricting the maximum age? That move would easily break half of the internet, so even if there is an intention to limit them, it will take place over a long period.

I don't really like the idea of sending cookies with every request, it increases the payload, especially for requests that have (close to) no response body. If you are really concerned, we can simply set the cookie to 2x the session lifetime and embed a timestamp/date which refreshes the cookie if it is older than 7 days or something. This makes sure that as long as there is a session, the cookie is still around and we only refresh the cookie once per week.